# RAD microbes
2025 RAD microbes boot camp

DATES: January 13th, 2025
LOCATION: O'Connor 5th floor event space | 6500 Main St., Houston, TX 77005

SUMMARY: Computational analyses of microbial genomes through genomic sequencing hold tremendous promise yet can become riddled with various sources of potential bias. The microbial genome analysis workshop is designed for graduate students, postdoctoral fellows, and investigators from diverse backgrounds and interests for clarity on how to effectively utilize ubiquitous computational pipelines for answering specific research questions. This hands-on workshop will cover end-to-end microbial genome analysis, discussing the pros and cons of decisions made during the process (library preparation, sequencing, bioinformatic tool selection, parameter settings, and interpretation of results).

TOPICS: Short and long-read sequencing, amplicon and isolate genome sequence, genome assembly and validation, functional annotation, phylogenomic analysis, and strain typing. This workshop will be taught by scientists with expertise in bioinformatics and analysis of clinical samples, allowing participants to get individualized training on how to accurately sample, sequence, and characterize microbial genomes.

## RAD Microbes Organizers and Instructors

### Dr. Rodrigo de Paula Baptista, PhD 

<img width="200" alt="image" src="https://github.com/treangenlab/radmicrobes/assets/28576450/5c815f66-0d01-4bab-87db-800f9b7a8ba3"> 

Rodrigo Baptista is a researcher with 14 years of expertise in the realms of eukaryotic pathogens and bioinformatics. His extensive experience includes a notable focus on parasite genomic data, particularly in the realms of Trypanosomatids and Apicomplexan parasites. Over the last few years, Baptista has cultivated specific proficiency in evolutionary genomics and the analysis of highly repetitive genomes and transcriptomes. His skill set encompasses tasks ranging from assembly and annotation to variant calling, and the identification and characterization of duplications and repetitive elements within these genomes. He has contributed significantly to the field by assembling and annotating numerous protozoan parasite genomes using various sequencing platforms, such as paired-end Illumina, Ion-Torrent short reads, PacBio, and Oxford Nanopore long-reads. These efforts have resulted in published data accessible to the scientific community. As a bioinformatician, he has established collaborations with research groups worldwide, working across diverse areas including protozoan parasites, prokaryotes, and vectors. Currently, at the Houston Methodist Research Institute, he leads the genomics enterprise within the Center for Infectious Diseases, focusing on employing genomics to characterize antibiotic resistance in clinical bacterial infections and eukaryotic pathogens. The primary goal is to mitigate the emergence and spread of resistant pathogens induced by treatment interventions.

### Dr. Blake Hanson, MS, PhD

<img width="200" alt="image" src="https://github.com/treangenlab/radmicrobes/assets/28576450/0aabd7fe-8836-4422-8f4e-bc335d661f87">

I am trained as an epidemiologist and have extensive experience in applying advanced genomic technologies and big-data analytical methods to investigate infectious diseases of public health importance. I am also an Assistant Professor in the Department of Internal Medicine, Division of Infectious Diseases, within the McGovern School of Medicine, and serve as the Associate Director for Microbial Genomics in the Center for Antimicrobial Resistance and Microbial Genomics (CARMiG). My laboratory uses a combination of existing and innovative laboratory techniques, and cutting-edge sequencing and bioinformatics to study infectious disease transmission and colonization, how microbial communities impact the development of disease, and how antimicrobial resistance develops and transmits through society. Current projects include: elucidating the importance of acquired antimicrobial resistance genes and mobile genetic elements in clinical outcomes, such as patient mortality and response to treatment, in Staphylococcus aureus causing bacteremia, co-circulating strains of carbapenem resistant Enterobacteriaceae (CRE), and vancomycin resistant enterococci (VRE); and interrogating the role of the microbiome in implanted medical device-associated infections, with a specific focus on breast implants placed following mastectomy due to cancer.

### Dr. Todd Treangen, PhD

<img width="200" alt="image" src="https://github.com/treangenlab/radmicrobes/assets/28576450/892207aa-f0e5-4299-8989-e03429ce5baf">

Todd Treangen is a Associate Professor in the Computer Science department at Rice University. He established his research group in the Computer Science department in July 2018. Prior to joining Rice, he was a Research Scientist at the Center for Bioinformatics and Computational Biology (CBCB) at the University of Maryland College Park, and previously a Principal Investigator at the National Biodefense Analysis and Countermeasures Center (NBACC). His primary research interests lie at the intersection of computer science and genomics, and his research group is focused on the development of novel computational methods and software tools with relevance to real-time monitoring of microbial community dynamics, infectious disease, and biothreats. Given the computational challenges presented by the metagenomic data deluge, coupled with the time-sensitive nature of problems specific to tracking pandemics and synthetic DNA screening, the Treangen lab strives to develop efficient and accurate computational solutions to emerging problems in these fields. Specifically, his research group focuses on the design, development, and implementation of novel algorithms, heuristics, and data structures to solve emerging computational research questions specific to biosecurity, infectious disease monitoring, and host-associated microbiome characterization. The Treangen lab is also dedicated to the dissemination and development of novel open-source bioinformatics methods, software, and pipelines, and to providing exciting “hands on” research opportunities to Rice undergraduates.

### Dr. Will Shropshire, PhD

<img width="200" alt="image" src="https://github.com/treangenlab/radmicrobes/assets/28576450/62d34750-c510-4993-9220-ee3b5de6cd37">

William Shropshire is a second year T32 Postdoctoral Fellow funded through the Gulf Coast Consortia/Keck Center’s Texas Medical Center Training Program in Antimicrobial Resistance (NIAID Grant No. T32 AI141349-05). He currently works at the University of Texas MD Anderson Cancer Center within the Department of Infectious Diseases, Infection Control, and Employee Health under the supervision of Dr. Samuel Shelburne. During the past six years, Dr. Shropshire has developed a strong skillset in computational biology and microbiology with the focus of his work translating genomic data into clinically impactful results. His pre- and post-doctoral work has spanned a broad spectrum of topics, ranging from the molecular mechanisms of antimicrobial resistance to the genomic epidemiology of infectious disease outbreaks. The focus of his current T32 project is to elucidate genomic and transcriptomic factors that contribute to the progressive development of carbapenem resistance within *Escherichia coli* causing invasive infections. Over the past three years, his research has documented the significant clinical impact of extended-spectrum beta-lactamase (ESBL) positive Enterobacterales and how ESBL encoding genes can amplify via mobile genetic elements upon initial exposures to beta-lactam drugs. There is increasing evidence that these beta-lactamase gene encoding amplifications can lead to heteroresistant populations, wherein only a subset of a bacterial population remains non-susceptible to treatment. His future goal is to leverage his T32 work into a K01 award in which he investigates the clinical impact of beta-lactam heteroresistance as well as other AMR survival strategies such as tolerance within high-risk Enterobacterales pathogens. His research work can be summarized [here](https://www.researchgate.net/profile/William-Shropshire).

### An Dinh

<img width="200" alt="image" src="https://github.com/treangenlab/radmicrobes/assets/153662215/e8e5f279-6442-4dc5-939a-9a407a6ec0fa">

An Dinh is a first year PhD student of epidemiology under Dr. Blake Hanson in the Center of Infectious Disease at the UTHealth School of Public Health.  For the past 8 years, he has worked with both short- and long- read sequencing technologies, producing data for large-scale genomic epidemiology studies of antimicrobial resistant pathogens.  His work experience includes laboratory automation, protocol development and optimization, high-throughput data processing, as well as more narrowly-focused small-scale projects. Studies have ranged from transcriptomics, metagenomics, and exploring the use of rapid sequencing of clinical samples for pathogen diagnoses and personalized medicine.  His interests include pushing new technologies and techniques to improve data resolution for AMR surveillance, optimizations of analyses pipelines for microbial genomics, and continuing to test and validate the utility of microbial sequencing in clinical spaces.

### Natalie Kokroko  

<img width="200" alt="image" src="https://github.com/treangenlab/radmicrobes/assets/761469/41e8dbd8-16ed-4fec-b612-de8f16613762">

Natalie Kokroko is a first year PhD student at Rice University in the Computer Science program. Her background is in Biomedical Engineering. Natalie worked in a research Institute at the University of Ghana (West African Centre for cell Biology of Infectious Pathogens) where she mainly did research and bioinformatics data analysis for the genomics and infectious disease laboratory. As a member of the Treangen lab, her research interest is to make use of computational tools and algorithms to interpret and analyze clinical and environmental microbiome data. Generally, Natalie is interested in Computational Biology, Bioinformatics, Genomics and Metagenomic data analysis. Her future goal is to be in academia and impart the knowledge and skills gained from her PhD to the next generation of scientists. 
