---
title: "3.2 - RADMicrobes_BootCamp"
author: "William Shropshire"
date: "`r Sys.Date()`"
output: html_document
---

### Core genome alignment with recombination masked using Gubbins for *K. pneumoniae* with a small subset (n=10) of isolates from CRACKLEII project

### Initial GGtree was inspired/borrowed from https://aschuerch.github.io/posts/2017-04-24-blog-post-1

```{r import libraries}
pacman::p_load(
        tidyverse,
        ape,
        ggtree,
        phytools,
        ggnewscale,
        ggtreeExtra,
        ggpubr,
        ggstar,
        scales, 
        ggplot2
)

```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Documents/GitHub/radmicrobes/session3/RScripts/")
```

### Note that the branch length scale with Gubbins is expressed in substitutions/genome

## Create basic iqtree and midroot the tree

```{r Create tree visualizations}
# Create tree for ML tree

# Create tree for ML tree
ml_tree <- read.tree("./../Files/phylogenetics/kpneumo.clean.full.aln.contree")
# midpoint root
ml_tree <- midpoint_root(ml_tree)

# Create ggtree basic plot
ggtree_ml <- ggtree(ml_tree) +
  geom_treescale(fontsize = 2, linesize = 0.5, family = "Arial Bold") +
  geom_tiplab(aes(label = label), size = 2, fontface = 2, family = "Helvetica")


# Create metadata set
df1 <- read.csv("./../Files/datasets/crackle2_kleborate_subset.csv")
df1$sra_id[df1$sra_id =="SRR12509439"] <- "Reference"
rownames(df1) <- df1$sra_id

p1 <- gheatmap(ggtree_ml, df1[, "ST", drop=FALSE], offset=0.0005, width=0.013, hjust = 1, colnames_angle = -45, colnames_position = "top", color = "white", font.size = 3, family="Arial Bold", custom_column_labels = c("ST"))+
  theme(legend.text=element_text(size=6, face="bold"),
        legend.title=element_text(size=10, face="bold"),
        legend.key.size = unit(0.5, 'cm'),
        legend.title.align=0.5) +
  ggplot2::ylim(-1.0, NA) + 
  coord_cartesian(clip = "off") + 
  scale_fill_manual(values = c("#ff0000", "#0000FF", "#5f0f40"), 
                    name = "Sequence Type",
                    breaks = c("ST307", "ST258", "ST244"))

p2 <- p1 + new_scale_fill()

p3 <- gheatmap(p2, df1[, "Bla_Carb_acquired", drop=FALSE], offset=0.00055, width=0.013, hjust = 1, colnames_angle = -45, colnames_position = "top", color = "white", font.size = 3, family="Arial Bold", custom_column_labels = c("Carb"))+
  theme(legend.text=element_text(size=6, face="bold"),
        legend.title=element_text(size=10, face="bold"),
        legend.key.size = unit(0.5, 'cm'),
        legend.title.align=0.5) +
  ggplot2::ylim(-1.0, NA) + 
  coord_cartesian(clip = "off") + 
  scale_fill_manual(values = c("#ff00e3", "#FFFFFFFF"), 
                    name = "KPC-2",
                    breaks = c("KPC-2", "-"), 
                    labels = c("Present", "Absent"))

ggsave("pneumo_ML_tree.svg",  
       p3,         
       device = "svg",       
       dpi = 300,             
       units = "cm",         
       height = 50,          
       width = 50)   

```

```{r clean up, echo=FALSE}
# Clear data
rm(list = ls())  

# Clear packages
pacman::p_unload(all)  

# Clear plots
graphics.off()  

# Clear console
cat("\014") 
```



